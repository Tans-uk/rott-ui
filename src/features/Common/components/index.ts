export * from './CommonItem'
export * from './CommonItemContainer'
