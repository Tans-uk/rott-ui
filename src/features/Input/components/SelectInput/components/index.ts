export * from './SelectInput'
export * from './SelectInputModalComponent'
