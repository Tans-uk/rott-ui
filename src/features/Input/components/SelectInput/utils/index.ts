export * from './listHeightNormalizer'
export * from './modalHeightPercentageNormalizer'
export * from './sortListBySearchPriority'
