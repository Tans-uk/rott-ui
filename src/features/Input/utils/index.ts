export * from './formatByDateMode'
export * from './inputNormalizer'
