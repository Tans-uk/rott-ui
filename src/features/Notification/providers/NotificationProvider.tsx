import {useCallback, useMemo, useState, type FC, type PropsWithChildren} from 'react'

import {
  NotificationContext,
  NotificationModel,
  notificationRef,
  ToasterNotificationProvider,
} from '..'
import {Item} from '../../Item'
import {useNotification} from '../hooks'

import {Toast} from 'react-native-toast-notifications'
import React from 'react'

let notifications = 0

const InitializeModalRef = () => {
  const notificationHook = useNotification()
  notificationRef.current = notificationHook

  return null
}

export const NotificationProvider: FC<PropsWithChildren> = ({children}) => {
  const [hasAnyNotification, setHasAnyNotification] = useState<boolean>(false)
  const increment = () => {
    notifications = notifications + 1

    if (notifications > 0) setHasAnyNotification(true)
  }

  const decrement = () => {
    if (notifications > 0) notifications = notifications - 1

    if (notifications === 0) setHasAnyNotification(false)
  }

  const baseNotification = ({title, description, variant}: NotificationModel) =>
    Toast.show(title, {
      type: variant,
      data: {title, description, variant},
      onClose: () => decrement(),
    })

  const show = useCallback(
    (notification: NotificationModel) => {
      increment()

      return baseNotification(notification)
    },
    [notifications]
  )
  const hide = useCallback(
    (id: string) => {
      decrement()

      return Toast.hide(id)
    },
    [notifications]
  )
  const hideAll = useCallback(() => {
    return Toast.hideAll()
  }, [])

  const success = useCallback(
    (title: string, description?: string) => {
      increment()

      return baseNotification({title, description, variant: 'success'})
    },
    [notifications]
  )
  const warning = useCallback(
    (title: string, description?: string) => {
      increment()

      return baseNotification({title, description, variant: 'warning'})
    },
    [notifications]
  )
  const danger = useCallback(
    (title: string, description?: string) => {
      increment()

      return baseNotification({title, description, variant: 'danger'})
    },
    [notifications]
  )
  const info = useCallback(
    (title: string, description?: string) => {
      increment()

      return baseNotification({title, description, variant: 'info'})
    },
    [notifications]
  )

  const contextValues = useMemo(() => {
    return {
      notifications,
      show,
      hide,
      hideAll,

      success,
      warning,
      danger,
      info,
    }
  }, [notifications, show, hide, hideAll, success, warning, danger, info, increment, decrement])

  return (
    <NotificationContext.Provider value={contextValues}>
      <InitializeModalRef />
      <Item flex={1} pointerEvents={hasAnyNotification ? 'box-only' : 'box-none'}>
        <ToasterNotificationProvider>{children}</ToasterNotificationProvider>
      </Item>
    </NotificationContext.Provider>
  )
}
