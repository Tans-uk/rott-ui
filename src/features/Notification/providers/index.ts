export * from './ToasterNotificationProvider'
export * from './NotificationProvider'
