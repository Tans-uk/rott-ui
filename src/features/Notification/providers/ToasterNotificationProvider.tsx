import {type FC, type PropsWithChildren} from 'react'

import {NotificationComponent} from '../components'

import {Icon, type IconTypes} from '@features/Icon'
import {type Variant} from '@models'

import {ToastProvider} from 'react-native-toast-notifications'

export const ToasterNotificationProvider: FC<PropsWithChildren> = ({children}) => {
  const notificationIcon = (name: IconTypes, variant: Variant) => {
    return <Icon name={name} width={20} height={20} mode='fill' variant={variant} />
  }

  return (
    <ToastProvider
      placement='top'
      duration={3000}
      animationType='slide-in'
      animationDuration={250}
      offset={50}
      offsetTop={30}
      offsetBottom={40}
      swipeEnabled={true}
      renderType={{
        success: ({onHide, data: {title, description}}) => (
          <NotificationComponent
            title={title}
            description={description}
            variantColor='rgba(72, 193, 181, 1)'
            iconElement={notificationIcon('CHECK_CIRCLE_FILL', 'success-notification')}
            onClose={onHide}
          />
        ),
        danger: ({onHide, data: {title, description}}) => (
          <NotificationComponent
            title={title}
            description={description}
            variantColor='rgba(209, 94, 83, 1)'
            iconElement={notificationIcon('INFORMATION_CIRCLE_FILL', 'danger-notification')}
            onClose={onHide}
          />
        ),
        warning: ({onHide, data: {title, description}}) => (
          <NotificationComponent
            title={title}
            description={description}
            variantColor='rgba(255, 173, 50, 1)'
            iconElement={notificationIcon('INFORMATION_CIRCLE_FILL', 'warning-notification')}
            onClose={onHide}
          />
        ),
        info: ({onHide, data: {title, description}}) => (
          <NotificationComponent
            title={title}
            description={description}
            variantColor='rgba(77, 175, 234, 1)'
            iconElement={notificationIcon('INFORMATION_CIRCLE_FILL', 'info-notification')}
            onClose={onHide}
          />
        ),
      }}>
      {children}
    </ToastProvider>
  )
}
