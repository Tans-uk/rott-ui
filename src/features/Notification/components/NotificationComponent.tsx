import {type FC, type JSX} from 'react'

import {themeConfig} from '../../../providers'
import {Item} from '../../Item'
import {Label} from '../../Label'
import {Pressable} from '../../Pressable'
import React from 'react'

interface NotificationMessageProps {
  title?: string
  description?: string | JSX.Element
  iconElement: React.ReactElement<any> | React.ReactNode
  onPress?: (() => void) | null | undefined
  onClose?: (() => void) | null | undefined
  variantColor: string
}

export const NotificationComponent: FC<NotificationMessageProps> = ({
  title,
  description,
  iconElement,
  variantColor,
  onPress,
  onClose,
}) => (
  <Pressable
    testID='notification-pressable-test-id'
    onPress={() => {
      !!onPress && onPress()
      !!onClose && onClose()
    }}
    marginBottom={24}
    relative>
    <Item
      absolute
      top={0}
      left={0}
      zIndex={99}
      width={4}
      height='100%'
      justifyContentCenter
      borderTopStartRadius={4}
      borderBottomStartRadius={4}
      backgroundColor={variantColor}
    />
    <Item
      row
      width={342}
      borderRadius={4}
      backgroundColor={themeConfig.colors['grey-900']}
      paddingVertical={16}
      paddingHorizontal={16}>
      <Item marginRight={8} marginTop={4}>
        {iconElement}
      </Item>

      <Item flex={1}>
        {title && (
          <Label
            fontSize='md'
            fontWeight={500}
            testID='notification-title-test-id'
            variant='white'
            lineBreakMode='tail'>
            {title}
          </Label>
        )}

        {description && (
          <Label
            fontSize='sm'
            fontWeight={500}
            testID='notification-desc-test-id'
            variant='white'
            marginTop={4}>
            {description}
          </Label>
        )}
      </Item>
    </Item>
  </Pressable>
)
