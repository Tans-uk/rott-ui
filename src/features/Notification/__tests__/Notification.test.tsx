import {fireEvent, render} from '../../../__tests__/utils/testUtils'
import {formatMessage} from '../../../libs'
import {NotificationComponent} from '../components'
import {NotificationProvider} from '../providers'

const mockProps = {
  title: formatMessage('TEST.WITH.PARAM', {testText: 'Test Title'}),
  description: formatMessage('TEST.WITH.PARAM', {testText: 'Test Description'}),
  mockOnClose: jest.fn(),
  mockOnPress: jest.fn(),
}

describe('Notification -> Custom Component', () => {
  const iconName = 'REMOVE'
  const testId = {
    notificationPressableTestId: 'notification-pressable-test-id',
    descriptionTestId: 'notification-desc-test-id',
    titleTestId: 'notification-title-test-id',
    blurTestId: 'notification-blur-test-id',
  }

  it('notification ilk render anında snapshot ile eşleşmeli', () => {
    const renderedNotification = render(
      <NotificationComponent
        variantColor={'info-notification'}
        iconElement={iconName}
        {...mockProps}
      />
    )

    expect(renderedNotification).toMatchSnapshot()
  })

  it('notification yokken ekranda blur gözükmemeli', () => {
    const {blurTestId} = testId
    const {queryByTestId} = render(<NotificationProvider />)

    const blurElement = queryByTestId(blurTestId)

    expect(blurElement).not.toBeOnTheScreen()
  })

  it('bildirime tıklandığında onPress fonksiyonları çağırılmalı', () => {
    const {notificationPressableTestId} = testId
    const {getByTestId} = render(
      <NotificationComponent
        variantColor={'info-notification'}
        iconElement={iconName}
        onPress={mockProps.mockOnPress}
        {...mockProps}
      />
    )

    const pressableElement = getByTestId(notificationPressableTestId)
    fireEvent.press(pressableElement)

    expect(mockProps.mockOnPress).toHaveBeenCalled()
    expect(mockProps.mockOnPress).toHaveBeenCalledTimes(1)
  })

  it('title verilen değerler ile ekranda gözükmeli', () => {
    const {title} = mockProps
    const {getByText} = render(
      <NotificationComponent
        variantColor={'info-notification'}
        iconElement={iconName}
        onClose={mockProps.mockOnClose}
        title={title}
      />
    )

    const titleElement = getByText(title)

    expect(titleElement).toBeOnTheScreen()
  })

  it('description verilen değerler ile ekranda gözükmeli', () => {
    const {description} = mockProps
    const {getByText} = render(
      <NotificationComponent
        variantColor={'info-notification'}
        iconElement={iconName}
        onClose={mockProps.mockOnClose}
        description={description}
      />
    )

    const descriptionElement = getByText(description)

    expect(descriptionElement).toBeOnTheScreen()
  })

  it('description verilmediği zaman ekranda gözükmemeli', () => {
    const {descriptionTestId} = testId
    const {queryByTestId} = render(
      <NotificationComponent
        variantColor={'info-notification'}
        iconElement={iconName}
        onClose={mockProps.mockOnClose}
      />
    )

    const descriptionElement = queryByTestId(descriptionTestId)

    expect(descriptionElement).not.toBeOnTheScreen()
  })
})
