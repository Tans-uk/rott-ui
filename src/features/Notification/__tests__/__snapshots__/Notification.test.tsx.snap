// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification -> Custom Component notification ilk render anında snapshot ile eşleşmeli 1`] = `
<View
  flex={1}
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <KeyboardProvider
    enabled={true}
  >
    <View
      flex={1}
      pointerEvents="box-none"
      style={
        {
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          animated={false}
          collapsable={false}
          focusable={true}
          marginBottom={24}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          relative={true}
          style={
            {
              "flex": 1,
              "flexDirection": "column",
              "marginBottom": 46,
              "position": "relative",
            }
          }
          testID="notification-pressable-test-id"
        >
          <View
            absolute={true}
            backgroundColor="info-notification"
            borderBottomStartRadius={4}
            borderTopStartRadius={4}
            height="100%"
            justifyContentCenter={true}
            left={0}
            style={
              {
                "backgroundColor": "rgba(77, 175, 234, 1)",
                "flex": undefined,
                "flexDirection": "column",
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": 8,
                "zIndex": 99,
              }
            }
            top={0}
            width={4}
            zIndex={99}
          />
          <View
            backgroundColor="rgba(34, 63, 70, 1)"
            borderRadius={4}
            paddingHorizontal={16}
            paddingVertical={16}
            style={
              {
                "backgroundColor": "rgba(34, 63, 70, 1)",
                "flex": undefined,
                "flexDirection": "row",
                "paddingHorizontal": 31,
                "paddingVertical": 31,
                "width": 658,
              }
            }
            width={342}
          >
            <View
              marginRight={8}
              marginTop={4}
              style={
                {
                  "flex": undefined,
                  "flexDirection": "column",
                  "marginRight": 16,
                  "marginTop": 8,
                }
              }
            >
              REMOVE
            </View>
            <View
              flex={1}
              style={
                {
                  "flex": 1,
                  "flexDirection": "column",
                }
              }
            >
              <Text
                lineBreakMode="tail"
                style={
                  {
                    "color": "rgba(255, 255, 255, 1)",
                    "fontFamily": "Markpro-Book",
                    "fontSize": 14,
                    "fontWeight": 500,
                    "textAlign": "auto",
                  }
                }
                testID="notification-title-test-id"
              >
                TEST WITH Test Title
              </Text>
              <Text
                marginTop={4}
                style={
                  {
                    "color": "rgba(255, 255, 255, 1)",
                    "fontFamily": "Markpro-Book",
                    "fontSize": 12,
                    "fontWeight": 500,
                    "marginTop": 8,
                    "textAlign": "auto",
                  }
                }
                testID="notification-desc-test-id"
              >
                TEST WITH Test Description
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "flexShrink": 1,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          [
            {
              "alignSelf": "center",
              "elevation": 999999,
              "flex": 0,
              "maxWidth": "100%",
              "position": "absolute",
              "zIndex": 999999,
            },
            {
              "flexDirection": "column-reverse",
              "justifyContent": "flex-start",
              "top": 30,
              "width": 750,
            },
          ]
        }
      >
        <View
          style={
            {
              "bottom": 0,
            }
          }
        >
          <RCTSafeAreaView />
        </View>
      </View>
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          [
            {
              "alignSelf": "center",
              "elevation": 999999,
              "flex": 0,
              "maxWidth": "100%",
              "position": "absolute",
              "zIndex": 999999,
            },
            {
              "bottom": 40,
              "flexDirection": "column",
              "justifyContent": "flex-end",
              "width": 750,
            },
          ]
        }
      >
        <View
          style={
            {
              "bottom": 0,
            }
          }
        >
          <RCTSafeAreaView />
        </View>
      </View>
    </View>
  </KeyboardProvider>
</View>
`;
